nameko_slack/rtm.py

```diff
@@ -2,73 +2,90 @@
 import re
 
 import eventlet
+from nameko.exceptions import ConfigurationError
 from nameko.extensions import Entrypoint, ProviderCollector, SharedExtension
 from slackclient import SlackClient
 
 
 EVENT_TYPE_MESSAGE = 'message'
 
 
-class SlackRTMClient(SharedExtension, ProviderCollector):
-    """
-    Slack Real Time Messaging API Client
+CONFIG_KEY = 'SLACK'
 
-    """
+
+DEFAULT_BOT_NAME = 'default'
+
+
+class SlackRTMClientManager(SharedExtension, ProviderCollector):
 
     def __init__(self):
 
-        super(SlackRTMClient, self).__init__()
+        super(SlackRTMClientManager, self).__init__()
 
         self.read_interval = 1
 
-        self.token = None
-
-        self._client = None
-        self._handlers = set()
+        self.clients = {}
 
     def setup(self):
-        config = self.container.config.get('SLACK', {})
-        self.token = config.get('TOKEN')
+
+        try:
+            config = self.container.config[CONFIG_KEY]
+        except KeyError:
+            raise ConfigurationError(
+                '`{}` config key not found'.format(CONFIG_KEY))
+
+        token = config.get('TOKEN')
+        clients = config.get('BOTS')
+        if token:
+            self.clients[DEFAULT_BOT_NAME] = SlackClient(token)
+        if clients:
+            for bot_name, token in clients.items():
+                self.clients[bot_name] = SlackClient(token)
+
+        if not self.clients:
+            raise ConfigurationError(
+                'At least one token must be provided in `{}` config'
+                .format(CONFIG_KEY))
 
     def start(self):
-        self._register_handlers()
-        self._connect()
-        self.container.spawn_managed_thread(self.run)
+        self.connect()
+        for bot_name, client in self.clients.items():
+            run = partial(self.run, bot_name, client)
+            self.container.spawn_managed_thread(run)
 
-    def run(self):
+    def run(self, bot_name, client):
         while True:
-            for event in self._client.rtm_read():
-                self.handle(event)
+            for event in client.rtm_read():
+                self.handle(bot_name, event)
             eventlet.sleep(self.read_interval)
 
-    def _register_handlers(self):
-        for provider in self._providers:
-            self._handlers.add(provider.handle_event)
-
-    def _connect(self):
-        self._client = SlackClient(self.token)
-        self._client.server.rtm_connect()
+    def connect(self):
+        for bot_name, client in self.clients.items():
+            client.server.rtm_connect()
 
-    def handle(self, event):
-        for handle_event in self._handlers:
-            handle_event(event)
+    def handle(self, bot_name, event):
+        for provider in self._providers:
+            if provider.bot_name == bot_name:
+                provider.handle_event(event)
```

Comment:
You are right, this uses the ProviderCollector providers set, which the mixin takes care of. It's a trade of between simplicity and efficiency. This could be revisited later on, with making the providers a dictionary.